From 9f13e99dd0fa4309ef7a893b75649cae16021159 Mon Sep 17 00:00:00 2001
From: Jean-Philippe Brucker <jean-philippe@linaro.org>
Date: Mon, 28 Feb 2022 17:05:39 +0000
Subject: [PATCH 70/97] KVM: arm64: pkvm: Add pkvm_create_hyp_device_mapping()

Add a function to map a MMIO region in the hypervisor and remove it from
the host. This allows hypervisor device drivers to reserve their regions
during setup.

Signed-off-by: Jean-Philippe Brucker <jean-philippe@linaro.org>
---
 arch/arm64/kvm/hyp/include/nvhe/mm.h |  1 +
 arch/arm64/kvm/hyp/nvhe/setup.c      | 19 +++++++++++++++++++
 2 files changed, 20 insertions(+)

diff --git a/arch/arm64/kvm/hyp/include/nvhe/mm.h b/arch/arm64/kvm/hyp/include/nvhe/mm.h
index 45b04bfee171..5764a289cf6e 100644
--- a/arch/arm64/kvm/hyp/include/nvhe/mm.h
+++ b/arch/arm64/kvm/hyp/include/nvhe/mm.h
@@ -25,5 +25,6 @@ int pkvm_create_mappings(void *from, void *to, enum kvm_pgtable_prot prot);
 int pkvm_create_mappings_locked(void *from, void *to, enum kvm_pgtable_prot prot);
 unsigned long __pkvm_create_private_mapping(phys_addr_t phys, size_t size,
 					    enum kvm_pgtable_prot prot);
+void __iomem *pkvm_create_hyp_device_mapping(u64 base, u64 size);
 
 #endif /* __KVM_HYP_MM_H */
diff --git a/arch/arm64/kvm/hyp/nvhe/setup.c b/arch/arm64/kvm/hyp/nvhe/setup.c
index c74c5bf21d77..c8db44d10c5c 100644
--- a/arch/arm64/kvm/hyp/nvhe/setup.c
+++ b/arch/arm64/kvm/hyp/nvhe/setup.c
@@ -265,6 +265,25 @@ static int fix_host_ownership(void)
 	return 0;
 }
 
+/* Map the MMIO region into the hypervisor and remove it from host */
+void __iomem *pkvm_create_hyp_device_mapping(u64 base, u64 size)
+{
+	void *virt;
+	int ret;
+
+	virt = (void *)__pkvm_create_private_mapping(base, size,
+						     PAGE_HYP_DEVICE);
+	if (IS_ERR(virt))
+		return 0;
+
+	/* lock not needed during setup */
+	ret = host_stage2_set_owner_locked(base, size, pkvm_hyp_id);
+	if (ret)
+		return ERR_PTR(ret);
+
+	return virt;
+}
+
 static int fix_hyp_pgtable_refcnt(void)
 {
 	struct kvm_pgtable_walker walker = {
-- 
2.34.1

